### PR TITLE
Add windows build version detection

### LINKED ISSUE #154796: I would like to see the actual Windows build in flutter doctor
### Use case

Multiple times when I created a issue and added the flutter doctor response I had to add the detail that I am using Windows 11 and not 10.

### Proposal

Check the Windows build number to detect the actual Windows version.

### PR DESCRIPTION
**Pull request by rekire**
With this PR flutter doctor will report the actual Windows version. This resolves #154796

The flutter doctor result is with this patch like this:
```
[√] Flutter (Channel stable, 3.24.2, on Microsoft Windows [Version 10.0.22631.4037], locale de-DE)
[√] Windows Version (Windows 11 - 23H2)
[√] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[√] Chrome - develop for the web
[√] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.8.3)
[√] Android Studio (version 2024.1)
[√] Connected device (3 available)
[√] Network resources

• No issues found!
```

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by rekire** [comment]
Found the issue which let the build fail. Until my fix is done I'll mark that PR as WIP

**Comment by loic-sharma** [comment]
Thanks for the contribution! Ideally we wouldn't need to maintain a list of Windows version manually. I wonder if instead we could use `Get-ComputerInfo` instead?

For example:

```
PS C:\Users\sharm> Get-ComputerInfo -Property OsName, OSDisplayVersion, WindowsVersion | Format-List


OsName           : Microsoft Windows 11 Pro
OSDisplayVersion : 23H2
WindowsVersion   : 2009
```

Perhaps we could try to read this. If this fails, we could fallback to `ver.exe` and today's implementation. What do you think of this idea?

**Comment by andrewkolos** [comment]
Agreed with Loic's comment. Given the thumbs up from @rekire on it, I assume this is planned route for this PR. I'll be marking it as a Draft to move it off the tool team's PR triage list. Please mark it as Ready for Review when it's time, and Loic and I will get notified 🙂.

**Comment by rekire** [comment]
FYI I'm still on the topic, but very busy

**Comment by rekire** [comment]
The build is green now, but I am not sure if the [timeout](https://ci.chromium.org/ui/p/flutter/builders/try/Windows%20module_test/42978/overview) can get more often with my change. The flutter doctor become a bit slower with my change. However I see no way to find a quicker API.

**Review by andrewkolos — CHANGES_REQUESTED** [review]
> The build is green now, but I am not sure if the [timeout](https://ci.chromium.org/ui/p/flutter/builders/try/Windows%20module_test/42978/overview) can get more often with my change. The flutter doctor become a bit slower with my change. However I see no way to find a quicker API.

Thanks for the continued work on this!

Testing this on my personal machine, the `Get-ComputerInfo` call seems to take ~2.2 seconds to run (`flutter doctor` in its entirety took 5s). I consider this performance regression too severe for a command as frequently run as `flutter doctor`.

From my own cursory research, it seems to best way to quickly determine whether the host machine is Windows 11 is to check that the build number is >= 22000[^1] (on a "Windows 10" machine). If we want to report "Windows 11" instead of Windows 10", I would use the build number to do so.

[^1]: https://stackoverflow.com/questions/69373447/is-there-an-official-way-to-detect-windows-11

**Comment by loic-sharma** [comment]
> The flutter doctor become a bit slower with my change.

Ah... I didn't realize my suggestion was not performant. Sorry about that.

> From my own cursory research, it seems to best way to quickly determine whether the host machine is Windows 11 is to check that the build number is >= 22000[1](https://github.com/flutter/flutter/pull/154797#user-content-fn-1-e6c6f124fadc88509857766d8fca4af2) (on a "Windows 10" machine). If we want to report "Windows 11" instead of Windows 10", I would use the build number to do so.

It looks like that's the same [approach `package:win32` uses](https://github.com/dart-windows/win32/blob/9728bc049bb8cb1953519363d79e0eba1432f4ff/examples/sysinfo.dart#L281-L286).

```dart
  final buildNumber = int.parse(getRegistryValue(
      HKEY_LOCAL_MACHINE,
      'SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion\\',
      'CurrentBuildNumber') as String);
  if (buildNumber >= 10240) print(' - Windows 10');
  if (buildNumber >= 22000) print(' - Windows 11');
```

Here's prior art in the Flutter tool that reads a Windows registry key: https://github.com/flutter/flutter/blob/beaf1fe68041c61299729b2096396583c0510d1e/packages/flutter_tools/lib/src/windows/visual_studio.dart#L470-L476

**Comment by rekire** [comment]
I am aware of that registry key, but it does not contain the exact edition e.g. "Windows 11 Pro". In the Registry is written "Windows 10 Pro" what makes IMHO absolutely no sense. Another option might be the WMI, but I guess that was is slow too.

**Comment by andrewkolos** [comment]
> I am aware of that registry key, but it does not contain the exact edition e.g. "Windows 11 Pro".

Just to make sure there is no misunderstanding: Loic is saying we could figure out whether we are running on Windows 10 or 11 by looking at the build number (specifically whether it is >= 22000).

>  Another option might be the WMI

If querying WMI is fast, I'd be happy to consider using that. I don't have access to my personal machine as I write this; but, on a weaker VM machine, `Measure-Command { wmic os get Caption | findstr /v Caption }` reports a time of 184ms, which I find acceptable (though still not ideal).

I'm fine with either option, so I'll leave it up to you as the implementer (cc @loic-sharma in case he has a stronger opinion).

**Comment by rekire** [comment]
I tried to use ffi and win32 for access that more quickly. I could not finish it yet, but am I allowed to add the win32 dependency for this use case?

**Comment by andrewkolos** [comment]
> I tried to use ffi and win32 for access that more quickly. I could not finish it yet, but am I allowed to add the win32 dependency for this use case?
> 
> Edit: Without the findstr part is is about 88ms on my computer. It works also fine on more "exotic" Windows versions like ﻿﻿﻿Microsoft Windows 10 IoT Enterprise LTSC.

I'd like us to avoid adding more dependencies to flutter tool whenever we can (though I suppose adding `win32` wouldn't be so bad since it only depends on `ffi`). If the powershell command would be simpler, I would prefer that.

**Comment by loic-sharma** [comment]
Quick correction - the Flutter project avoids third-party dependencies and we recently-ish decided to remove our dependency on `package:win32`. For example: https://github.com/flutter/packages/pull/7073

> Per https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#dependencies, we generally do not want third-party dependencies. `path_provider` in particular is a key part of the ecosystem (well over 1,000 packages depend directly on it, and transitively it's probably several times that at least), and thus the maintenance and security considerations are particularly acute.

We'll need to write the FFI code by hand instead of using that package.

**Comment by rekire** [comment]
Okay I use now a mixed way. I ran into a memory corruption so I stopped using the WMI directly. I execute now the command and take the missing information (the DisplayVersion and the ReleaseId) from the registry. I hope that I'll get the build green again, so that you can test it.

**Comment by andrewkolos** [comment]
> We'll need to write the FFI code by hand instead of using that package.

I suspect that maintaining native code for the sake of printing more a more user-friendly OS name in `flutter doctor` is not worth it. I don't want this code doing anything more complicated than a simple command-line utility, but I am not even sure we need to do that—doesn't the version string the tool gets from dart:io already have everything we need need? For example, `10.0.22631.4037` tells us that the kernel version is 10 and that the build number is `22631`. From this, the tool could determine that it is running on Windows 11, no?

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (nate-thegrate) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:140]:
```suggestion
      final String? output = getProcessesResult.stdout as String?;
```

**Review by nate-thegrate — COMMENTED** [review]
Giving some advice per a recent [Discord thread](https://discord.com/channels/608014603317936148/608018585025118217/1291374162002182270):

```
   info • Local variables should be final • packages/flutter_tools/lib/src/windows/windows_version_validator.dart:140:7 • prefer_final_locals
   info • Missing type annotation • packages/flutter_tools/lib/src/windows/windows_version_validator.dart:149:12 • always_specify_types
   info • The type name 'KEY_VALUE_BASIC_INFORMATION' isn't an UpperCamelCase identifier • packages/flutter_tools/lib/src/windows/windows_version_validator.dart:187:12 • camel_case_types
   info • The variable name 'TitleIndex' isn't a lowerCamelCase identifier • packages/flutter_tools/lib/src/windows/windows_version_validator.dart:189:16 • non_constant_identifier_names
   info • The variable name 'Type' isn't a lowerCamelCase identifier • packages/flutter_tools/lib/src/windows/windows_version_validator.dart:191:16 • non_constant_identifier_names
   info • The variable name 'NameLength' isn't a lowerCamelCase identifier • packages/flutter_tools/lib/src/windows/windows_version_validator.dart:193:16 • non_constant_identifier_names
   info • The variable name 'DataLength' isn't a lowerCamelCase identifier • packages/flutter_tools/lib/src/windows/windows_version_validator.dart:195:16 • non_constant_identifier_names
   info • Missing type annotation • packages/flutter_tools/lib/src/windows/windows_version_validator.dart:205:3 • always_specify_types
   info • The variable name 'RegOpenKeyEx' isn't a lowerCamelCase identifier • packages/flutter_tools/lib/src/windows/windows_version_validator.dart:206:26 • non_constant_identifier_names
   info • The variable name 'RegQueryValueEx' isn't a lowerCamelCase identifier • packages/flutter_tools/lib/src/windows/windows_version_validator.dart:207:29 • non_constant_identifier_names
   info • The variable name 'RegCloseKey' isn't a lowerCamelCase identifier • packages/flutter_tools/lib/src/windows/windows_version_validator.dart:208:25 • non_constant_identifier_names
   info • Missing type annotation • packages/flutter_tools/lib/src/windows/windows_version_validator.dart:214:3 • always_specify_types
   info • Missing type annotation • packages/flutter_tools/lib/src/windows/windows_version_validator.dart:220:3 • always_specify_types
   info • Missing type annotation • packages/flutter_tools/lib/src/windows/windows_version_validator.dart:225:3 • always_specify_types
   info • Redundant initialization to 'null' • packages/flutter_tools/lib/src/windows/windows_version_validator.dart:230:11 • avoid_init_to_null
   info • Missing a newline at the end of the file • packages/flutter_tools/lib/src/windows/windows_version_validator.dart:246:2 • eol_at_end_of_file

16 issues found. (ran in 180.4s)
```

I'll give a suggestion for the one at the top—see if you can figure out the rest!

**Comment by rekire** [comment]
_@andrewkolos wrote:_
> [...] doesn't the version string the tool gets from dart:io already have everything we need need? For example, `10.0.22631.4037` tells us that the kernel version is `10.0` and that the build number is `22631`. From this, the tool could determine that it is running on Windows 11, no?

Technically correct. However since Windows 11 and Windows 10 share the same kernel version 10.0, it is worth to differentiate the build numbers. My first simple idea way to have a list to provide a "readable" name. The power shell way was too slow. I took then the wmic command, which is quicker. I thought it would be nice to have the patch level like "23H2" too. Instead of starting another process I thought it would be more efficient to use FFI for that. Do you disagree?

As I already wrote in the initial comment I don't want to mention that I am using Windows 11 and not Windows 10 or up.

I am not absolutely sure what differences are there between the different Windows editions, but I know for Example that Windows 10 IoT Enterprise LTSC has a Kiosk mode which other versions do not have. This is why I think the exact version can be very relevant.

@nate-thegrate
Thank you that helps me a lot to understand the problem.

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (nate-thegrate) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:243]: (entirely optional suggestion, feel free to ignore)

```suggestion
  calloc
    ..free(subKeyPtr)
    ..free(valueNamePtr)
    ..free(hSubKeyPtr)
    ..free(dataSizePtr)
    ..free(typePtr)
    ..free(dataPtr);
```

**Review by nate-thegrate — COMMENTED** [review]
Unfortunately I'm not familiar with the **flutter_tools** architecture, so I'm only providing feedback with regards to Dart/Flutter code style.

Overall this LGTM, though I'll defer to other reviewers to give an approval.

**Review by andrewkolos — CHANGES_REQUESTED** [review]
> Technically correct. However since Windows 11 and Windows 10 share the same kernel version 10.0, it is worth to differentiate the build numbers. My first simple idea way to have a list to provide a "readable" name.

Right, but _zero_ additional shell processes or native syscalls are required for us to differentiate between Windows 10 and 11.  The flutter tool, as it exists today, already has access to the build number (e.g. `Version 10.0.22631.4169`, which appears in the output of `flutter doctor -v` on my machine). From this _alone_, we see the build number, `22631`, is greater than `22000`. This identifies that we are running on a Windows 11 host. If the build number were less than `22000`, we would know that we're running on a Windows 10 host.

I'm unable to tell whether my earlier explanation of this was confusing, or if you were simply re-summarizing this thread.

> I thought it would be nice to have the patch level like "23H2" too.


> Instead of starting another process I thought it would be more efficient to use FFI for that. Do you disagree?

Using FFI is likely more efficient. However, this comes at the cost of a significant amount of nontrivial code. For me, maintainability of the flutter tool is significantly more important than its performance. I did some local testing, and reading `ReleaseId` and `DisplayVersion` synchronously from the system registry using the `reg` command in Dart took roughly 65 milliseconds. The FFI method in this PR took 6 milliseconds (in both cases I did not include a `wmic` call, which is by far the slowest thing). I would rather pay the 60 millisecond penalty than introduce FFI code into the tool. If 60 milliseconds is too great a cost, than I think this feature is not worth implementing.

**Comment by rekire** [comment]
Just to summarize your change request:
I should not try to get the Windows edition? Instead I should try to get the Windows Version from the build number. I should prefer executing shell commands in order to get much slower, but mush less code.

IMHO the FFI code is worth to keep it. I think it is more or less write once code which does not really needs to be maintained, such basic Win-API trend not to change at all. However I agree with you the paradigm that no code is good code.

So what's next? Removing the FFI code, take the patch level via the reg command and parse the build number to get the Windows version and drop the (new) support for the edition?

**Comment by andrewkolos** [comment]
> I should not try to get the Windows edition? 

You are welcome to. However, if you do, I'm asking for it to be done without FFI for the sake of those reading/maintaining the code. The `reg` command should be sufficient. (Sorry for the confusion, the main point of my previous comment was to point out that FFI isn't required to differentiate between Windows 10 and Windows 11, but it seems you already shared this understanding with me).

One other thing that scares me with using FFI is what could happen if the native code invocations resulted in a crash for whatever reason. I fear that this would end the flutter tool process (rather than a Dart error/exception being raised) and thus we wouldn't get any crash reports.

**Comment by andrewkolos** [comment]
I've added a couple of tests and done some light refactoring. Feel free to revert or revise these commits as desired. I still have some concerns to work through here, so I'll provide another review soon.

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:102]: If the `reg` call fails, this will print something like `Windows 11 or higher, null, null`. I added a failing test to cover this behavior. Perhaps we could do `<String>[caption, details.displayVersion, details.releaseId].join(', ')` instead? With this, we should just see `Windows 11 or higher`.

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:177]: We may want to log the error here with `printTrace` just in case someone reports a bug in this feature and does not want to debug it themselves.

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (rekire) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:102]: That's a good idea. I was not aware that the join method excludes `null`s

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/windows/windows_version_validator.dart:198]: Perhaps we should log the error here. (See earlier comment https://github.com/flutter/flutter/pull/154797/files#r1805859210 for more)

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (rekire) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:102]: Actually it does not work. when I run `print(<String?>['hallo', null, 'world'].join(' '));` I get the output "hello null world", so those need to filtered out first

Comment (andrewkolos) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart]: > Actually it does not work. when I run `print(<String?>['hallo', null, 'world'].join(' '));` I get the output "hello null world", so those need to filtered out first

Oh yeah, I forgot that part. Sorry 😓

**Comment by rekire** [comment]
@andrewkolos what are the next steps? I think all open suggestions are applied and the latest ci run was green.

**Review by andrewkolos — APPROVED** [review]
LGTM. Sorry for the initial misdirection regarding whether or not to use ffi here.

**Review by bkonyi — COMMENTED** [review]
Generally LGTM, but we should add some documentation.

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (rekire) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:96]: Good that you point it out, actually it is wrong and should be 22000. That is the lowest Windows 11 build number as documented by Microsoft: https://learn.microsoft.com/en-us/windows/release-health/windows11-release-information. For Windows 10 is there a similar list: https://learn.microsoft.com/en-us/windows/release-health/release-information

File: packages/flutter_tools/lib/src/windows/windows_version_validator.dart
Comment (rekire) [reply packages/flutter_tools/lib/src/windows/windows_version_validator.dart:165]: done

**Comment by bkonyi** [comment]
Awesome! Thanks for the documentation!

This LGTM overall, but it looks like you'll need to address the analyzer failure first.

**Comment by rekire** [comment]
@bkonyi done, but what is that new failed test?

**Comment by bkonyi** [comment]
> @bkonyi done, but what is that new failed test?

Just some internal testing that failed to run correctly. This is a relatively safe change that shouldn't impact anything covered by that test so I've bypassed it.

Thanks for the contribution!

**Comment by rekire** [comment]
@bkonyi how does the cherry-pick mechanism work? Do you cherry-pick it or do I need to create a new PR for that? I am not sure how it works.

**Comment by bkonyi** [comment]
> @bkonyi how does the cherry-pick mechanism work? Do you cherry-pick it or do I need to create a new PR for that? I am not sure how it works.

I don't think this would be a candidate for a cherry-pick into the upcoming stable release since it's not fixing a critical bug. Unfortunately, you'll have to wait until the next beta release or stable (sometime early next year) if you're not able to switch to the master channel.

**Comment by rekire** [comment]
That's fine. So the next time the current beta gets the stable status, the current main is getting forked was beta?

**Comment by jmagman** [comment]
You can read about how the channels (branches) work in https://github.com/flutter/flutter/blob/master/docs/releases/Flutter-build-release-channels.md#flutters-channels

> We branch from master for a new beta release at the beginning of the month, usually the first Wednesday. This includes a branch for Dart, the engine and the framework. These branches are then "stabilized" for the next couple of weeks, meaning we accept [cherrypick](https://github.com/flutter/flutter/blob/master/docs/releases/Flutter-Cherrypick-Process.md) requests for high impact issues. Once a quarter, the beta branch lives on to become the next stable branch, as detailed below.

**Pull request merged**
